src/openrct2/core/OrcaStream.hpp

```diff
@@ -0,0 +1,685 @@
+                        auto& el = vec.emplace_back();
+                        f(el);
+                        NextArrayElement();
+                    }
+                    EndArray();
+                }
+                else
+                {
+                    BeginArray();
+                    for (auto& el : vec)
+                    {
+                        f(el);
+                        NextArrayElement();
+                    }
+                    EndArray();
+                }
+            }
+
+            template<typename TArr, size_t TArrSize, typename TFunc> void ReadWriteArray(TArr (&arr)[TArrSize], TFunc f)
+            {
+                auto& arr2 = *(reinterpret_cast<std::array<TArr, TArrSize>*>(arr));
+                ReadWriteArray(arr2, f);
+            }
+
+            template<typename TArr, size_t TArrSize, typename TFunc>
+            void ReadWriteArray(std::array<TArr, TArrSize>& arr, TFunc f)
+            {
+                if (_mode == Mode::READING)
+                {
+                    auto count = BeginArray();
+                    for (auto& el : arr)
+                    {
+                        el = {};
+                    }
+                    for (size_t i = 0; i < count; i++)
+                    {
+                        if (i < TArrSize)
+                        {
+                            f(arr[i]);
+                        }
+                        NextArrayElement();
+                    }
+                    EndArray();
+                }
+                else
+                {
+                    BeginArray();
+                    for (auto& el : arr)
+                    {
+                        if (f(el))
+                        {
+                            NextArrayElement();
+                        }
+                    }
+                    EndArray();
+                }
+            }
+
+            template<typename T> void Ignore()
+            {
+                T value{};
+                ReadWrite(value);
+            }
+
+        private:
+            void ReadBuffer(void* dst, size_t len)
+            {
+                _buffer.Read(dst, len);
+            }
+
+            void WriteBuffer(const void* buffer, size_t len)
+            {
+                _buffer.Write(buffer, len);
+            }
+
+            template<typename T, typename = std::enable_if<std::is_integral<T>::value>> T ReadInteger()
+            {
+                if constexpr (sizeof(T) > 4)
+                {
+                    if (std::is_signed<T>())
+                    {
+                        int64_t raw{};
+                        Read(&raw, sizeof(raw));
+                        return static_cast<T>(raw);
+                    }
+
+                    uint64_t raw{};
+                    Read(&raw, sizeof(raw));
+                    return static_cast<T>(raw);
+                }
+                else
+                {
+                    if constexpr (std::is_signed<T>())
```

Comment:
Can be combined to reduce indention by one level.

Reply:
You mean combine the >4 and the signed check? That would lead to harder to follow code.

Reply:
No, I mean instead of `else { if (...) { ... } }` you can do `else if (...) { ... }`.

Reply:
yeah thats what i mean it would be harder to follow